### PR TITLE
p2p: remove adjusted time

### LINKED ISSUE #4521: AddTimeData will never update nTimeOffset past 199 samples
Reading the function AddTimeData, I came to notice two things:
- The structure vTimeOffsets contains up to 200 elements, after which any new element added to it will not increase its size, replacing the oldest element.
- The condition to update nTimeOffset includes checking whether the number of elements in vTimeOffsets is odd, which will never happen after there are 200 elements.

This will lead to nTimeOffset not updating past 199 calls to AddTimeData, remaining forever stuck in whatever value it had then. I don't believe this is the inteded behaviour. I would suggest removing the oddness check, since vTimeOffsets is capable of returning the median value of an even list according to the median method of the CMedianFilter type.

In current master, the offending code is in src/timedata.cpp, line 52, columns 34-64:

``` c++
if (vTimeOffsets.size() >= 5 && vTimeOffsets.size() % 2 == 1)
```

This should better be:

``` c++
if (vTimeOffsets.size() >= 5)
```

### PR DESCRIPTION
**Pull request by fanquake**
After recent refactoring, the changes required to remove adjusted time are now quite straight forward. This PR removes the notion of adjusted time, along with the `-maxtimeadjustment` option. It doesn't do any other cleanup.

Opening for discussion of gotchas / brainstorming / concept (N)ACKs.

1 question from a reviewer was whether we should keep some sort of warning around for if our clock differs too much from our peers. Although it's unclear how useful that might be.

Commits can be cleaned up / sqaushed. Missing release notes etc.
Would close #4521.

### CODE REVIEW

**Comment by naumenkogs** [comment]
Concept ACK, light code review ACK.

File: src/rpc/net.cpp [hunk 1]
```diff
@@ -638,7 +636,6 @@ static RPCHelpMan getnetworkinfo()
     if (node.peerman) {
         obj.pushKV("localrelay", !node.peerman->IgnoresIncomingTxs());
     }
-    obj.pushKV("timeoffset",    GetTimeOffset());
```
Comment (MarcoFalke) [inline src/rpc/net.cpp:641]: Maybe keep this field and instead return the  median offset of the currently connected outbounds or so? See also https://github.com/bitcoin/bitcoin/pull/24606#issuecomment-1107815177

**Review by maflcko — COMMENTED** [review]
Concept ACK, but there should be an easy way for users to figure out if their clock is wrong, ideally at startup or early after startup. It wouldn't be the best UX if they found out by going out of consensus or by seeing their mined blocks be dropped.

File: src/net_processing.cpp [hunk 2]
```diff
@@ -2988,13 +2987,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
...
                   remoteAddr);
 
-        int64_t nTimeOffset = nTime - GetTime();
-        pfrom.nTimeOffset = nTimeOffset;
-        if (!pfrom.IsInboundConn()) {
-            // Don't use timedata samples from inbound peers to make it
-            // harder for others to tamper with our adjusted time.
-            AddTimeData(pfrom.addr, nTimeOffset);
-        }
+        pfrom.nTimeOffset = nTime - GetTime();
```
Comment (ajtowns) [inline src/net_processing.cpp:2990]: Perhaps:

```c++
if (!pfrom.IsInboundConn() && std::abs(pfrom.nTimeOffset) >= 1800) {
    ++m_bad_non_inbound_time_offset;
    if (m_bad_non_inbound_time_offset == 3) {
        SetMiscWarning("Please check your computer's date and time are correct");
    }
}
``` 

(ie, add a counter to `PeerManagerImpl`, and on the third outbound connection that has a time offset of 30mins or more, set a warning that will show up in `getblockchaininfo`, `getnetworkinfo`, `getmininginfo` until restart)

**Comment by dergoegge** [comment]
Concept ACK

**Review by ariard — COMMENTED** [review]
By removing the adjustement of our local clock time on the median of offsets from peers clocks, we're lowering the robustness of our operational time against time shifting attacks where an adversary shifts the local time of the host NTP client backward/forward. Before, this change even if the clock drifts, it's at least corrected in the bound of `DEFAULT_MAX_TIME_ADJUSTEMENT`. After, this change if the clock is controlled by malicious timeservers, I don't think there is corrective action or even warnings issued towards the operators.

With most of Linux distributions, the local time at a client is determined based on the clock readings received from timeservers from the [NTP pool](https://www.ntppool.org/en/), an open set of public servers, segmented in geographical zones for load-balancing. It should be noted the [policy of this pool](https://www.ntppool.org/en/use.html) recommends one to distrust time provided if high reliability is sought. An adversary could either fulfill the pool with crappy servers or exploit one of the unpatched [vulnerabilities](https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=ntpd) as the high-stratum servers are likely randomly maintained due to being run by volunteers. 

For block consensus validation, we're relying on clock time accuracy w.r.t honest network of peers, a block timestamp too far in the future _from the viewpoint of the local clock_ will be rejected (L3472 in `ContextualCheckBlockHeader()`). A successful time shifting attack against your time could fork you off from the network and from then escalate to a [time-dilation attack ](https://arxiv.org/pdf/2006.01418.pdf) variant against your Lightning node.

While I think this is valuable to overhaul or remove `GetAdjustedTime()`, I would prefer first that we introduce a warning system based on discrepancies between local clock and outbound peers ones, or allocate more thoughts if we can come up with better mitigations against NTP attacks at the application-level.

cc @TheBlueMatt @naumenkogs with whom I had discussions in the past on how fucking up with your NTP syncing can compromise your Lightning node funds security.

**Comment by MarcoFalke** [comment]
> Before, this change even if the clock drifts, it's at least corrected in the bound of DEFAULT_MAX_TIME_ADJUSTEMENT.

No correction will happen if your time drifts and you only ever cycle through at most 4 distinct honest outbound connections. Also, no correction will happen if your time drifts after you made more than `BITCOIN_TIMEDATA_MAX_SAMPLES/2+1` distinct honest outbound connections. Instead of a false promise that adjusted time will correct an incorrect system clock, it would be better for the system operator to correct the system clock.

I don't see a way that adjusted time can be helping some without also harming others. Every instance where it provides a "fix" for some users, it opens the attack surface for users that are not affected by the problem. For example, if an attacker can influence enough (not all) outgoing connections of your node, they can likely get you out of consensus with a perfectly synced system clock. This also breaks the assumption that a single honest connection would be sufficient for you to stay in consensus.

**Comment by ariard** [comment]
> No correction will happen if your time drifts and you only ever cycle through at most 4 distinct honest outbound connections. Also, no correction will happen if your time drifts after you made more than BITCOIN_TIMEDATA_MAX_SAMPLES/2+1 distinct honest outbound connections. Instead of a false promise that adjusted time will correct an incorrect system clock, it would be better for the system operator to correct the system clock.

From my understanding, there is at least a correction brought to your system clock once you have at most 4 distinct honest outbound connections and before you have made more than BITCOIN_TIMEDATA_MAX_SAMPLES/2+1, which I think made my point correct that the existence of a correction hardens the attack game for an adversary targeting your NTP syncing, even minimally. Noting this correction is featherweight if you're facing an adversary with BGP capabilities, I don't deny it.

I would agree a better protection would be effectively for the system operator to correct the system clock, a good question being if we can achieve this at the bitcoind-level with a warning system potentially based on peers clocks, or we should recommend high-stakes users to be their own stratum 0 servers, and directly get time from GPS or a non-Internet time source.

> For example, if an attacker can influence enough (not all) outgoing connections of your node, they can likely get you out of consensus with a perfectly synced system clock. 

Assuming you're starting with a perfectly synced system clock, as we're bounded with `DEFAULT_MAX_TIME_ADJUSTEMENT`, of an effective value of 70min, the attacker won't be able to make you reject block due to seen timestamp being too far in the future. From my comprehension, current adjusted time does not break the assumption that a single honest connection would be sufficient to stay in consensus.

**Comment by MarcoFalke** [comment]
They could delay your time by 70min and then mine blocks 2h ahead (or advance the time of another miner by 70min), which would make you lag on consensus and throw away your hashrate (if you have any) for as long as the latest block is too far ahead.

File: src/net_processing.cpp (hunk 2 above)
Comment (ajtowns) [reply src/net_processing.cpp:2990]: Alternatively, we regularly do new feeler and blocks-only connections, perhaps we could keep a rolling average of `abs(peer.nTimeOffset)` and report that in `getnetworkinfo`?

**Comment by darosior** [comment]
Concept ACK.

**Comment by fanquake** [comment]
Rebased on master. @ajtowns putting aside the current comments, any thoughts on including changes like this in bitcoin-inquisition?

**Comment by ajtowns** [comment]
> Rebased on master. @ajtowns putting aside the current comments, any thoughts on including changes like this in bitcoin-inquisition?

No objection to it, but not sure why it would be easier to do it that way than just merge into master?

**Pull request closed without merging**